### ISSUE TITLE
Nav Unification: Menu Item click - event tracking

### ISSUE THREAD

**Issue by getdave** [Automattic/wp-calypso#50847]
Prior to nav-unification each menu item had an explicit tracking event

eg https://github.com/Automattic/wp-calypso/blob/729bc1d55f73cbac61a2b68e03576d727c4474fe/client/my-sites/sidebar/index.jsx#L183

We need to find a way to track clicks in the dynamic menu served from the API.

**Comment by tjcafferkey** [comment]
Just some initial thoughts on this

* This tracking will need implementing in Calypso and WP Admin to cover the scenarios where the user is in either one or the other.
  - Does JP have active tracking? I can see [this](https://github.com/Automattic/jetpack/blob/master/projects/packages/tracking/src/js/tracks-ajax.js) but haven't seen anything to suggest its being used, or even if its the same system used in Calypso?
  - If it is being used it looks like it can rely on data attributes. Would [this (nav_menu_link_attributes)](https://developer.wordpress.org/reference/hooks/nav_menu_link_attributes/) be an option?
* We can't use user facing labels due to i18n. Is using the slug an option?
  - Potential problem with slugs is they contain site names (e.g. `/posts/your-site.wordpress.com`) for Calypso slugs so we would need a reliable way to handle this.
* We could end up tracking interactions with third-party plugin menu items. Is this a good or bad thing?

**Comment by cpapazoglou** [comment]
> We can't use user facing labels due to i18n. Is using the slug an option?

We can use the slug after striping the domain. 

> We could end up tracking interactions with third-party plugin menu items. Is this a good or bad thing?

Sounds like something we could really benefit from ;-)!

> This tracking will need implementing in Calypso and WP Admin to cover the scenarios where the user is in either one or the other.

Looking in tracks funnels, I can see there aren't a vast amount of funnels using `calypso_mysites_sidebar_` in Steps.

![](https://cln.sh/OQCgYJ+)

Perhaps, instead of trying to find a WP Admin solution for tracking, we could focus on fixing the funnels already used. For example:  
- a User is in WP Admin
- clicks Themes (which would trigger `calypso_mysites_sidebar_themes_clicked`)
- lands in Calypso with `https://wordpress.com/themes/[domain]?**track-user-click**
- we can add a function in Calypso themes page that triggers `calypso_mysites_sidebar_themes_clicked` if `track-user-click` is in params.

**Comment by cpapazoglou** [comment]
I can see the following tracking events currently used in Funnels

- `calypso_mysites_sidebar_view_site_clicked`
- `calypso_mysites_sidebar_customer_home_clicked`
- `calypso_mysites_sidebar_customize_clicked`
- `calypso_mysites_sidebar_themes_clicked`
- `calypso_mysites_sidebar_wpadmin_clicked`
- `calypso_mysites_sidebar_plan_clicked`

**Comment by tjcafferkey** [comment]
>we can add a function in Calypso themes page that triggers calypso_mysites_sidebar_themes_clicked if track-user-click is in params.

We can do this for links that take users from WP Admin > Calypso but it won't work for WP Admin > WP Admin (as you mentioned above).

We may also struggle to reliably track a click on a Calypso > WP Admin menu item with the same event name as a Calypso > Calypso menu item if we are using the slug.

For example if the user has the dashboard toggle enabled do we want to track clicks to Posts > All Posts from Home (Calypso) with the same event name regardless of whether the destination is WP Admin or Calypso? As the slugs are different (`https://wordpress.com/posts/sitename.wordpress.com` or `https://sitename.wordpress.com/wp-admin/edit.php`) it creates added complexity by using the slug to track this intent.

Another approach (if it's possible) is that no matter what language the menu item is in we always translate it back to English and turn that into an event name. This way regardless of menu item destination (whether it's going to Calypso or WP Admin) it's always the same event that gets tracked.

**Comment by tjcafferkey** [comment]
### The issue

Problem 1: We are no longer tracking menu clicks in Calypso
Problem 2: We don't currently have a method to reliably track menu clicks in WP Admin

These are going to be treated separately. I'm going to focus on fixing the regression which is highlighted in _Problem 1_.

**The problem with reliable events in Calypso**

* Labels change based on language settings.
* Slugs change based on dashboard settings (Users with the toggle on will get WP Admin slugs).
* We need to map the 'new' events to the events that are being actively used in funnels.

### The solution

Adding a menu ID into the menu response will bring some consistency and reliability to events. Regardless of language or dashboard settings this ID should not change for the menu item.

_We need to bear in mind that this solution will eventually be needed in WP Admin._

* Add a menu item into the menu response.
* To maintain old events we can make a table to map menu ids to event names that are currently being used.

**Comment by cpapazoglou** [comment]
You are right @tjcafferkey, I haven't thought about users navigating from WP Admin to WP Admin. In that case your solution makes sense. 

I guess we can abuse the `$GLOBALS['menu']` here 
https://github.com/Automattic/jetpack/blob/01bf918ccb0322a2b2fff2048e671ddf8fc42d2d/projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php#L47

and by looping it and adding the identifiers in the menu items we wish. In that context we both know the Domain and the WP Admin link preference so checking for the slug should be doable

![](https://cln.sh/GisKUc+)

**Comment by cpapazoglou** [comment]
cc @mmtr who is now an expert in WordPress menu api!

**Comment by mmtr** [comment]
Thanks for the ping. I briefly discussed this with @tjcafferkey yesterday and agreed that adding a menu/submenu identifier in the API response is the best way forward.

However, I think the global `$submenu` doesn't have any identifier that we can re-use so we would probably need to autogenerate one based on maybe its slug.

Anyway, the goal is to keep tracking the same events that we were tracking before, and all these identifiers will probably not match the current event names. To solve this, we may need to create a map in Calypso that translates the API identifiers to the event names.

**Comment by tjcafferkey** [comment]
Discussion moved to here paYJgx-1ua-p2

**Comment by Copons** [comment]
The discussion (paYJgx-1ua-p2) was picked up again and we agreed on this:

- As initially suggested by @tjcafferkey, we will use page views events instead of sidebar menu item clicks.
  Page view events already exist, are more reliable, and don't need any special handling or future maintenance.

- We'll therefore need to deprecate and remove sidebar menu item click events.

- On the other hand, we'll need to make sure that JITM nudges will fire the same events in both Calypso and wp-admin.

**Comment by rcrdortiz** [comment]
I've checked that both Calypso and wp-admin are using the same tracking event names when dealing with JITM nudges.
